observabilityclient/v1/deploy.py

```diff
@@ -0,0 +1,132 @@
+#   Copyright 2022 Red Hat, Inc.
+#
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may
+#   not use this file except in compliance with the License. You may obtain
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+
+import os
+import requests
+import sys
+import yaml
+
+from osc_lib.i18n import _
+
+from observabilityclient.v1 import base
+from observabilityclient.utils import runner
+
+
+INVENTORY = os.path.join(base.OBSWRKDIR, 'openstack-inventory.yaml')
+ENDPOINTS = os.path.join(base.OBSWRKDIR, 'scrape-endpoints.yaml')
+STACKRC = os.path.join(base.OBSWRKDIR, 'stackrc')
+
+
+def _curl(host: dict, port: int, timeout: int = 1) -> str:
+    """Returns scraping endpoint URL if it is reachable
+    otherwise returns None."""
+    url = f'http://{host["ip"]}:{port}/metrics'
+    try:
+        r = requests.get(url, timeout=1)
+        if r.status_code != 200:
+            url = None
+        r.close()
+    except requests.exceptions.ConnectionError:
+        url = None
+    return url
+
+
+class Discover(base.ObservabilityBaseCommand):
+    """Generate Ansible inventory file and scrapable enpoints list file."""
+
+    def get_parser(self, prog_name):
+        parser = super().get_parser(prog_name)
+        parser.add_argument(
+            '--scrape',
+            action='append',
+            default=['collectd/9666'],
+            help=_("Service/Port of scrape endpoint to check on nodes")
+        )
+        parser.add_argument(
+            '--stack-name',
+            default='overcloud',
+            help=_("Overcloud stack name for which inventory file should "
+                   "be generated")
+        )
+        return parser
+
+    def take_action(self, parsed_args):
+        # discover undercloud and overcloud nodes
+        rc, out, err = self._execute(
+            'tripleo-ansible-inventory '
```

Comment:
I am having trouble getting discovery to work on my Wallaby cloud:
```
(undercloud) [stack@undercloud osp-observability-ansible]$ openstack observability discover --scrape collectd/9103                                                             [644/1884]
/usr/lib64/python3.6/site-packages/_yaml/__init__.py:23: DeprecationWarning: The _yaml extension module is now located at yaml._yaml and its location is subject to change.  To use the L
ibYAML-based parser and emitter, import from `yaml`: `from yaml import CLoader as Loader, CDumper as Dumper`.                                                                           
  DeprecationWarning                                                                                                                                                                  
[WARNING]: Unable to parse /home/stack/.osp-observability/openstack-inventory.yaml as an inventory source                                                                                
[WARNING]: No inventory was parsed, only implicit localhost is available                                                                
Discovered following scraping endpoints:                                                                                                                                                 
{}  
```

So I tried running this step manually and this is what happened:
```
(undercloud) [stack@undercloud .osp-observability]$ tripleo-ansible-inventory --static-yaml-inventory ./openstack-inventory.yaml --stack overcloud
DEPRECATED - This script has been deprecated.Use the inventory file tripleo-ansible-inventory.yaml located under ~/tripleo-deploy/<stack> for a Standalone/Undercloud deployment or overcloud-deploy/<stack> for the Overcloud. 
To generate an inventory file, use the playbook in tripleo-ansible: cli-config-download.yaml

(undercloud) [stack@undercloud .osp-observability]$ ls
scrape-endpoints.yaml
```

"Deprecated" may not be the correct word, because it looks to me like it doesn't even work at all.

Luckily the inventory it suggests is actually there:
```
(undercloud) [stack@undercloud .osp-observability]]$ egrep 'overcloud-.*:$' ~/overcloud-deploy/overcloud/tripleo-ansible-inventory.yaml 
    overcloud-controller-0:
    overcloud-controller-1:
    overcloud-controller-2:
    overcloud-compute-0:
```
So I copied it into position, and tried again:
```
(undercloud) [stack@undercloud .osp-observability]$ cp ~/overcloud-deploy/overcloud/tripleo-ansible-inventory.yaml openstack-inventory.yaml                            
                 
(undercloud) [stack@undercloud .osp-observability]$ openstack observability discover --scrape collectd/9103                                                                             
/usr/lib64/python3.6/site-packages/_yaml/__init__.py:23: DeprecationWarning: The _yaml extension module is now located at yaml._yaml and its location is subject to change.  To use the LibYAML-based parser and emitter, import from `yaml`: `from yaml import CLoader as Loader, CDumper as Dumper`.                                                                           
  DeprecationWarning
('cannot represent an object', 'overcloud-compute-0')
```

Reply:
I patched the source as per my suggestion below (reverting the change I erroneously asked for)

```
(undercloud) [stack@undercloud python-observabilityclient]$ git diff
diff --git a/observabilityclient/v1/deploy.py b/observabilityclient/v1/deploy.py
index b62fba9..537634f 100644
--- a/observabilityclient/v1/deploy.py
+++ b/observabilityclient/v1/deploy.py
@@ -86,9 +86,7 @@ class Discover(base.ObservabilityBaseCommand):
                           f'{name} at port {port}', end='')
                 node = _curl(host, port, timeout=1)
                 if node:
-                    endpoints.setdefault(host["host"], dict()).setdefault(
-                        service.strip(), []
-                    ).append(node)
+                    endpoints.setdefault(service.strip(), []).append(node)
                 if parsed_args.dev:
                     print(' [success]' if node else ' [failure]')
         data = yaml.safe_dump(endpoints, default_flow_style=False)
```

Now it works again:

```
(undercloud) [stack@undercloud .osp-observability]$ openstack observability discover --scrape collectd/9103                                                                             
/usr/lib64/python3.6/site-packages/_yaml/__init__.py:23: DeprecationWarning: The _yaml extension module is now located at yaml._yaml and its location is subject to change.  To use the L
ibYAML-based parser and emitter, import from `yaml`: `from yaml import CLoader as Loader, CDumper as Dumper`.                                                                           
  DeprecationWarning
Discovered following scraping endpoints:
collectd:
- http://192.168.24.27:9103/metrics
- http://192.168.24.19:9103/metrics
- http://192.168.24.29:9103/metrics
- http://192.168.24.18:9103/metrics
```

So the upshot:
- The inventory generation appears broken in Wallaby (likely OSP 17 as well)
- Please undo the changes made to accommodate Prometheus being installed to every host - not only was that a silly request from a confused co-worker (me), they don't appear to work anyways

Reply:
Hmm, interesting. I was developing this patch on OSP17 environment, but noted. Will add fallback to the suggested inventory file. Thanks for another try Chris!